resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/hadoopsteps/HadoopKerberosKeytabResolverStep.scala

```diff
@@ -0,0 +1,171 @@
+  * assumes that the job user has either specified a principal and keytab or ran
+  * $kinit before running spark-submit. With a TGT stored locally, by running
+  * UGI.getCurrentUser you are able to obtain the current user, alternatively
+  * you can run UGI.loginUserFromKeytabAndReturnUGI and by running .doAs run
+  * as the logged into user instead of the current user. With the Job User principal
+  * you then retrieve the delegation token from the NameNode and store values in
+  * DelegationToken. Lastly, the class puts the data into a secret. All this is
+  * appended to the current HadoopSpec which in turn will append to the current
+  * DriverSpec.
+  */
+private[spark] class HadoopKerberosKeytabResolverStep(
+  submissionSparkConf: SparkConf,
+  maybePrincipal: Option[String],
+  maybeKeytab: Option[File],
+  hadoopUGI: HadoopUGIUtil) extends HadoopConfigurationStep with Logging{
+    private var originalCredentials: Credentials = _
+    private var dfs : FileSystem = _
+    private var renewer: String = _
+    private var credentials: Credentials = _
+    private var tokens: Iterable[Token[_ <: TokenIdentifier]] = _
+
+    override def configureContainers(hadoopConfigSpec: HadoopConfigSpec): HadoopConfigSpec = {
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(submissionSparkConf)
+    logDebug(s"Hadoop Configuration: ${hadoopConf.toString}")
+    if (hadoopUGI.isSecurityEnabled) logError("Hadoop not configuration with Kerberos")
+    val maybeJobUserUGI =
+      for {
+        principal <- maybePrincipal
+        keytab <- maybeKeytab
+      } yield {
+        // Not necessary with [Spark-16742]
+        // Reliant on [Spark-20328] for changing to YARN principal
+        submissionSparkConf.set("spark.yarn.principal", principal)
+        submissionSparkConf.set("spark.yarn.keytab", keytab.toURI.toString)
+        logDebug("Logged into KDC with keytab using Job User UGI")
+        hadoopUGI.loginUserFromKeytabAndReturnUGI(
+          principal,
+          keytab.toURI.toString)
+      }
+    // In the case that keytab is not specified we will read from Local Ticket Cache
+    val jobUserUGI = maybeJobUserUGI.getOrElse(hadoopUGI.getCurrentUser)
+    // It is necessary to run as jobUserUGI because logged in user != Current User
+    jobUserUGI.doAs(new PrivilegedExceptionAction[Void] {
+      override def run(): Void = {
+        logDebug(s"Retrieved Job User UGI: $jobUserUGI")
+        originalCredentials = jobUserUGI.getCredentials
+        logDebug(s"Original tokens: ${originalCredentials.toString}")
+        logDebug(s"All tokens: ${originalCredentials.getAllTokens}")
+        logDebug(s"All secret keys: ${originalCredentials.getAllSecretKeys}")
+        dfs = FileSystem.get(hadoopConf)
+        // This is not necessary with [Spark-20328] since we would be using
+        // Spark core providers to handle delegation token renewal
+        renewer = jobUserUGI.getShortUserName
```

Comment:
What is the value add in making this a config key?

Reply:
We plan to run a token refresh server in a separate pod for extending the lifetime of jobs tokens, in case spark jobs run for days and weeks and their original tokens expire. The refresh server would use a system user Kerberos principal that may be different from the job user principal. This renewer should match the refresh server's principal name. Otherwise, the namenode will reject the renew request. Since we do not know which system user principal will be used, hence the need to make this config option.